singer-commons/src/main/java/com/pinterest/singer/loggingaudit/client/AuditEventKafkaSender.java

```diff
@@ -0,0 +1,420 @@
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditHeaders;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditEvent;
+import com.pinterest.singer.loggingaudit.thrift.LoggingAuditStage;
+import com.pinterest.singer.loggingaudit.thrift.configuration.KafkaSenderConfig;
+import com.pinterest.singer.metrics.OpenTsdbMetricConverter;
+
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.thrift.TException;
+import org.apache.thrift.TSerializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ *  LoggingAuditEventSender implementations that dequeue LoggingAuditEvent and send it to Kafka.
+ *
+ *  Each instance of this class should be executed in its own thread.
+ */
+public class AuditEventKafkaSender implements LoggingAuditEventSender {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuditEventKafkaSender.class);
+
+  private static final int MAX_RETRIES_FOR_SELECTION_RANDOM_PARTITION = 10;
+
+  private static final int PARTITIONS_REFRESH_INTERVAL_IN_SECONDS = 300;
+
+  /**
+   *  When sender send audit events to Kafka,  it chooses a random partition and if it fails, it
+   *  will choose another GOOD partition, this retry will happen at most numOfPartitionsToTrySending
+   *  times before dropping the event. Note that, this is different from the retry handled by
+   *  kafka client
+   *  library when sending an event to a certain partition.
+   */
+  private static final int NUM_OF_PARTITIONS_TO_TRY_SENDING = 3;
+
+  /**
+   *  maximum time to wait when sender tries to dequeue events before returning null.
+   */
+  private static final int DEQUEUE_WAIT_IN_SECONDS = 30;
+
+
+  /**
+   *   when gracefully shutting down the Sender, the calling thread sleeps for some time and let
+   *   the sender to send out audit events left in the queue if there are any.
+   *   stopGracePeriodInSeconds is the maximum time reserved and
+   *   stopGracePeriodInSeconds / THREAD_SLEEP_IN_SECONDS is the total rounds the calling thread
+   *   will sleep.
+   */
+  private static final int THREAD_SLEEP_IN_SECONDS = 10;
+
+  /**
+   *  when gracefully shutting down the Sender, this field specifies maximum time for main thread
+   *  to wait, in order to let the sender send out audit events left in the queue if there are any.
+   */
+  private int stopGracePeriodInSeconds = 300;
```

Comment:
can we lower the default value to 15 seconds?    300 seconds is too long. we don't want to be in the blocking path during java service restart.

Reply:
it makes sense.  while the total grace period is configurable, making the default to be 30 seconds is good in terms of not blocking service restart.  I made the change in the update of the PR